# CIAN.ru

In [1]:
import requests
import re
from bs4 import BeautifulSoup
import pandas as pd
import time

In [2]:
def html_stripper(text):
    return re.sub('<[^<]+?>', '', str(text))

### Добавляем ссылки на все округа в массив

In [3]:
CAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=13&district%5B1%5D=14&district%5B2%5D=15&district%5B3%5D=16&district%5B4%5D=17&district%5B5%5D=18&district%5B6%5D=19&district%5B7%5D=20&district%5B8%5D=21&district%5B9%5D=22&engine_version=2&offer_type=flat&p={}&room1=1&room2=1&room3=1&room4=1&room5=1&room6=1'
SAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=23&district%5B10%5D=33&district%5B11%5D=34&district%5B12%5D=35&district%5B13%5D=36&district%5B14%5D=37&district%5B15%5D=38&district%5B1%5D=24&district%5B2%5D=25&district%5B3%5D=26&district%5B4%5D=27&district%5B5%5D=28&district%5B6%5D=29&district%5B7%5D=30&district%5B8%5D=31&district%5B9%5D=32&engine_version=2&offer_type=flat&room1=1&room2=1'
SVAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=39&district%5B10%5D=49&district%5B11%5D=50&district%5B12%5D=51&district%5B13%5D=52&district%5B14%5D=53&district%5B15%5D=54&district%5B16%5D=55&district%5B1%5D=40&district%5B2%5D=41&district%5B3%5D=42&district%5B4%5D=43&district%5B5%5D=44&district%5B6%5D=45&district%5B7%5D=46&district%5B8%5D=47&district%5B9%5D=48&engine_version=2&offer_type=flat&room1=1&room2=1'
VAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=56&district%5B10%5D=66&district%5B11%5D=67&district%5B12%5D=68&district%5B13%5D=69&district%5B14%5D=70&district%5B15%5D=71&district%5B1%5D=57&district%5B2%5D=58&district%5B3%5D=59&district%5B4%5D=60&district%5B5%5D=61&district%5B6%5D=62&district%5B7%5D=63&district%5B8%5D=64&district%5B9%5D=65&engine_version=2&offer_type=flat&room1=1&room2=1'
UVAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=72&district%5B10%5D=82&district%5B11%5D=83&district%5B1%5D=73&district%5B2%5D=74&district%5B3%5D=75&district%5B4%5D=76&district%5B5%5D=77&district%5B6%5D=78&district%5B7%5D=79&district%5B8%5D=80&district%5B9%5D=81&engine_version=2&offer_type=flat&room1=1&room2=1'
UAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=84&district%5B10%5D=94&district%5B11%5D=95&district%5B12%5D=96&district%5B13%5D=97&district%5B14%5D=98&district%5B15%5D=99&district%5B1%5D=85&district%5B2%5D=86&district%5B3%5D=87&district%5B4%5D=88&district%5B5%5D=89&district%5B6%5D=90&district%5B7%5D=91&district%5B8%5D=92&district%5B9%5D=93&engine_version=2&offer_type=flat&room1=1&room2=1'
UZAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=100&district%5B10%5D=110&district%5B11%5D=111&district%5B1%5D=101&district%5B2%5D=102&district%5B3%5D=103&district%5B4%5D=104&district%5B5%5D=105&district%5B6%5D=106&district%5B7%5D=107&district%5B8%5D=108&district%5B9%5D=109&engine_version=2&offer_type=flat&room1=1&room2=1'
ZAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=112&district%5B10%5D=122&district%5B11%5D=123&district%5B12%5D=124&district%5B13%5D=348&district%5B14%5D=349&district%5B15%5D=350&district%5B1%5D=113&district%5B2%5D=114&district%5B3%5D=115&district%5B4%5D=116&district%5B5%5D=117&district%5B6%5D=118&district%5B7%5D=119&district%5B8%5D=120&district%5B9%5D=121&engine_version=2&offer_type=flat&room1=1&room2=1'
SZAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=125&district%5B1%5D=126&district%5B2%5D=127&district%5B3%5D=128&district%5B4%5D=129&district%5B5%5D=130&district%5B6%5D=131&district%5B7%5D=132&engine_version=2&offer_type=flat&room1=1&room2=1'
ZelAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=152&district%5B1%5D=153&district%5B2%5D=154&district%5B3%5D=355&district%5B4%5D=356&district%5B5%5D=357&district%5B6%5D=358&engine_version=2&offer_type=flat&room1=1&room2=1'
NAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=327&district%5B10%5D=337&district%5B1%5D=328&district%5B2%5D=329&district%5B3%5D=330&district%5B4%5D=331&district%5B5%5D=332&district%5B6%5D=333&district%5B7%5D=334&district%5B8%5D=335&district%5B9%5D=336&engine_version=2&offer_type=flat&room1=1&room2=1'
TAO = 'http://www.cian.ru/cat.php?deal_type=sale&district%5B0%5D=338&district%5B1%5D=339&district%5B2%5D=340&district%5B3%5D=341&district%5B4%5D=342&district%5B5%5D=343&district%5B6%5D=344&district%5B7%5D=345&district%5B8%5D=346&district%5B9%5D=347&engine_version=2&offer_type=flat&room1=1&room2=1'
districts = [CAO,SAO,SVAO,VAO,UVAO,UAO,UZAO,ZAO,SZAO, ZelAO,NAO,TAO ]

### В цикле по округам и страницам в них качаем ссылки на квартиры

In [5]:
links = []
for district in districts:
    for page in range(1, 30):
        page_url =  district.format(page)

        search_page = requests.get(page_url)
        search_page = search_page.content
        search_page = BeautifulSoup(search_page, 'lxml')

        flat_urls = search_page.findAll('div', attrs = {'ng-class':"{'serp-item_removed': offer.remove.state, 'serp-item_popup-opened': isPopupOpen}"})
        flat_urls = re.split('http://www.cian.ru/sale/flat/|/" ng-class="', str(flat_urls))

        for link in flat_urls:
            if link.isdigit():
                links.append(link)

In [6]:
len(links)

8990

### Попробуем перейти на первую

In [7]:
flat_url = 'http://www.cian.ru/sale/flat/' + str(links[0]) + '/'
#flat_url = 'http://www.cian.ru/sale/flat/150531912/' 
flat_page = requests.get(flat_url)
flat_page = flat_page.content
flat_page = BeautifulSoup(flat_page, 'lxml')
flatStats = {'District':0}

# Цена

In [8]:
def getPrice(flat_page):
    price = flat_page.find('div', attrs={'class':'object_descr_price'})
    price = re.split('<div>|руб|\W', str(price))
    price = "".join([i for i in price if i.isdigit()][-3:])
    return int(price)

# Координаты

In [9]:
def getCoords(flat_page):
    coords = flat_page.find('div', attrs={'class':'map_info_button_extend'}).contents[1]
    coords = re.split('&amp|center=|%2C', str(coords))
    coords_list = []
    for item in coords:
        if item[0].isdigit():
            coords_list.append(item)
    lat = float(coords_list[0])
    lon = float(coords_list[1])
    return lat, lon

# Число комнат

In [10]:
def getRoom(flat_page):
    rooms = flat_page.find('div', attrs={'class':'object_descr_title'})
    rooms = html_stripper(rooms)
    room_number = ''
    for i in re.split('-|\n', rooms):
        if 'комн' in i:
            break
        else:
            room_number += i
    room_number = "".join(room_number.split())
    return room_number

# Таблица

In [11]:
table = flat_page.find('table', attrs = {'class':'object_descr_props'})

In [12]:
table = html_stripper(table)

# Общая площадь

In [13]:
def getTotsp(flat_page):
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    Totsp_n=re.split('Общая площадь|Площадь комнат', table)[1]
    Totsp = ''
    for i in re.split(':|\xa0|-|\n', Totsp_n):
        if 'м2' in i:
            break
        else:
            Totsp += i
    Totsp = "".join(Totsp.split())
    return Totsp 

# Жилая площадь

In [14]:
def getLivesp(flat_page):
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    Livesp_n=re.split('Жилая площадь|Площадь кухни', table)[1]
    Livesp = ''
    for i in re.split(':|\xa0|-|\n', Livesp_n):
        if 'м2' in i:
            break
        else:
            Livesp += i
    Livesp = "".join(Livesp.split())
    return Livesp

# Площадь кухни

In [15]:
def getKitsp(flat_page):
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    Kitsp_n1=re.split('Площадь кухни|Совмещенных санузлов', table)[1]
    Kitsp1 = ''
    for i in re.split(':|\xa0|-|\n', Kitsp_n1):
        if 'м2' in i:
            break
        else:
            Kitsp1 += i
    Kitsp1 = "".join(Kitsp1.split())
    Kitsp_n2=re.split('Площадь кухни|Раздельных санузлов', table)[1]
    Kitsp2 = ''
    for i in re.split(':|\xa0|-|\n', Kitsp_n2):
        if 'м2' in i:
            break
        else:
            Kitsp2 += i
    Kitsp2 = "".join(Kitsp2.split())
    Kitsp_n3=re.split('Площадь кухни|Санузел', table)[1]
    Kitsp3 = ''
    for i in re.split(':|\xa0|-|\n', Kitsp_n3):
        if 'м2' in i:
            break
        else:
            Kitsp3 += i
    Kitsp3 = "".join(Kitsp3.split())
    if len(Kitsp1)<5:
        Kitsp=Kitsp1
    elif len(Kitsp3)<5:
        Kitsp=Kitsp3
    else:
        Kitsp=Kitsp2  
    return Kitsp 

# Количество этажей

In [16]:
def getNfloors(flat_page):    
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    Nfloors_n=re.split('Этаж|Тип дома', table)[1]
    Nfloors1 = re.split(':|\xa0|-', Nfloors_n)
    L=Nfloors1[-1]
    Nfloors3=re.split('\n', L)
    Nfloors=Nfloors3[0]
    return Nfloors

# Наличие балкона

In [17]:
def getBal(flat_page): 
    table = flat_page.find('table', attrs = {'class':'object_descr_props'}) 
    table = html_stripper(table) 
    Bal_n=re.split('Балкон|Лифт', table)[1] 
    Bal1=re.split(':|\n', Bal_n)[2] 
    Bal1 
    if Bal1=='нет': 
        Bal=0
    elif len(Bal1)>1: 
        Bal=1    
    else: 
        Bal=None 
    return Bal 

# Расстояние до метро в минутах

In [18]:
def getMetrdist(flat_page): 
    Metro = flat_page.find('div', attrs={'class':'object_descr_metro'}) 
    Metro = html_stripper(Metro) 
    Metr_dist = '' 
    for i in re.split('\n|\n', Metro): 
        if 'мин.' in i: 
            break 
        else: 
            Metr_dist = i 
    Metr_dist = "".join(Metr_dist.split()) 
    return Metr_dist 

# Пешком до метро или на транспорте

In [19]:
def getWalk (flat_page):   
    Metro = flat_page.find('div', attrs={'class':'object_descr_metro'}) 
    Metro = html_stripper(Metro) 
    Metro1 = html_stripper(Metro)
    Metro1
    if len(re.split('\n|\n', Metro1))>=5:
        N_n=re.split('\n|\n', Metro1)[-5]
        if '                    пешком' == N_n: 
            Walk=1 
        else: 
            Walk = 0
    else:
        Walk=None
    return Walk

# Материал строения

In [20]:
def getBrick(flat_page):  
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    New_n=re.split('Тип дома|Тип продажи', table)[1]
    New1 = ''
    for i in re.split(':|\xa0|-|\n', New_n):
        if 'м2' in i:
            break
        else:
            New1 += i
    New1 = "".join(New1.split())
    New_2=re.split(',', New1)
    New_3=re.split('1|2|3|4|5|6|7|8|9|0|,', New1)
    if (len(New_2)>=2) and (len(New_2)<=3):
        Brick1 = New_2[1]
        if Brick1=='монолит':
            Brick=1
        elif Brick1=='монолитный':
            Brick=1
        elif Brick1=='кирпичный':
            Brick=1   
        elif Brick1=='жб':
            Brick=1
        elif Brick1=='кирпичномонолитный':
            Brick=1
        elif New_3[1]=='монолитВысотапотолков':
            Brick=1
        elif New_3[1]=='монолитныйВысотапотолков':
            Brick=1
        elif New_3[1]=='кирпичныйВысотапотолков':
            Brick=1   
        elif New_3[1]=='жбВысотапотолков':
            Brick=1
        elif New_3[1]=='кирпичномонолитныйВысотапотолков':
            Brick=1
        else:
            Brick=0
    else:
        Brick=None
    return Brick

# Первичный рынок/вторичка

In [21]:
def getNew(flat_page):  
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    New_n=re.split('Тип дома|Тип продажи', table)[1]
    New1 = ''
    for i in re.split(':|\xa0|-|\n', New_n):
        if 'м2' in i:
            break
        else:
            New1 += i
    New1 = "".join(New1.split())
    New_2=re.split(',', New1)
    New3 =New_2[0]
    if New3=='вторичка':
        New=0
    elif New3=='первичный рынок':
        New=1
    else:
        New=None
    return New

### Телефон

In [22]:
def getTel(flat_page):  
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    Tel_n=re.split('Телефон|Вид из окна', table)[1]
    Tel1=re.split(':|\n', Tel_n)[2]
    Tel1
    if Tel1=='нет':
        Tel=0
    elif Tel1=='да':
        Tel=1
    elif Tel1=='есть':
        Tel=1
    elif len(Tel1)>1:
        Tel=1
    else:
        Tel=None
    return Tel

# Номер этажа

In [23]:
def getFloor(flat_page):
    table = flat_page.find('table', attrs = {'class':'object_descr_props'})
    table = html_stripper(table)
    Floor_n=re.split('Этаж|Тип дома', table)[1]
    Floor = ''
    for i in re.split(':|\xa0|-|\n', Floor_n):
        if '/' in i:
            break
        else:
            Floor += i
    Floor = "".join(Floor.split())
    return Floor 

# Расстояние от центра в км

In [24]:
import math 
from math import sqrt
flatStats['lat'], flatStats['lon'] = getCoords(flat_page) 
x_cent = 37.609218 
y_cent = 55.753559 
gr_to_km = 6372.795/360.0 
coordinates = (flatStats['lat'],flatStats['lon'])
def dist(coordinates): 
    return(sqrt((coordinates[1]-x_cent)**2+(coordinates[0]-y_cent)**2)*gr_to_km) 

### Теперь делаем цикл и собираем все наши функции вместе

In [26]:
d = {}
i = 0
t=[]
for j in range(0, len(links)):
    t.append(links[j])
for link in t:
    flat_url = 'http://www.cian.ru/sale/flat/' + str(link) + '/' 
    flat_page = requests.get(flat_url) 
    flat_page = flat_page.content 
    flat_page = BeautifulSoup(flat_page, 'lxml')
    flatStats = {'District':0} 
    flatStats['Rooms'] = getRoom(flat_page)
    flatStats['Price'] = getPrice(flat_page) 
    flatStats['Totsp'] = getPrice(flat_page) 
    flatStats['lat'], flatStats['lon'] = getCoords(flat_page) 
    flatStats['Totsp'] = getTotsp(flat_page)
    flatStats['Livesp'] = getLivesp(flat_page)
    flatStats['Kitsp'] = getKitsp(flat_page)
    flatStats['Floor'] = getFloor(flat_page)
    flatStats['Nfloors'] = getNfloors(flat_page)
    flatStats['Dist'] = dist(coordinates) 
    flatStats['Bal'] = getBal(flat_page)
    flatStats['Tel'] = getTel(flat_page)
    flatStats['New'] = getNew(flat_page)
    flatStats['Brick'] = getBrick(flat_page)
    flatStats['Metrdist'] = getMetrdist(flat_page)
    flatStats['Walk'] = getWalk(flat_page)
    flatStats['Livesp'] = getLivesp(flat_page)
    flatStats['Kitsp'] = getKitsp(flat_page)
    flatStats['Floor'] = getFloor(flat_page)
    flatStats['Nfloors'] = getNfloors(flat_page)
    flatStats['Dist'] = dist(coordinates) 
    flatStats['Bal'] = getBal(flat_page)
    flatStats['Tel'] = getTel(flat_page)
    flatStats['New'] = getNew(flat_page)
    flatStats['Brick'] = getBrick(flat_page)
    flatStats['Metrdist'] = getMetrdist(flat_page)
    flatStats['Walk'] = getWalk(flat_page)
    i+=1
    flatStats['N'] = i
    d[i] = flatStats

ChunkedEncodingError: ("Connection broken: ConnectionAbortedError(10053, 'Программа на вашем хост-компьютере разорвала установленное подключение', None, 10053, None)", ConnectionAbortedError(10053, 'Программа на вашем хост-компьютере разорвала установленное подключение', None, 10053, None))

### Выкачалось 4184 ссылки из 8990. У человека, с которым вместе писали код всё сработало. Возожно, дело в компьютере. Полный датасет для анализа можно будет взять у него. 

In [27]:
df =pd.DataFrame.from_dict(d,orient='index')
df.to_csv('out.csv')